Subscription model modular architecture

ahmedali8 opened the pull request.

ahmedali8 committed:
build: add cre8ors and update remappings
Refs: #7

ahmedali8 committed:
feat: add Admin access control

Utilize cre8ors admin role

ahmedali8 committed:
feat: add Minter abstract contract access control

Inherits Admin to create onlyMinter to be used for controlled subscription

ahmedali8 committed:
feat: add Payment System abstract contract

Handles native and erc20 tokens for subscription

Refs: #8

ahmedali8 committed:
feat: add Subscription interface
docs: add natspec doc

Refs: #9

ahmedali8 committed:
feat: add ERC20 Transfer Helper library for consistency

ahmedali8 committed:
fix: updateSubscription should be payable

ahmedali8 committed:
refactor: change fallback architecure to interface architecture
feat: inherit PaymentSystem and ISubscription
feat: implement erc20 payment methods

ahmedali8 committed:
refactor: delete MockERC721

ahmedali8 committed:
feat: add mock nft contract

Mock NFT contract mimicks cre8ors contract and overrides ownerOf function to call subscription contract externally and validate subscription status
Setter functions must be only access controlled by onlyAdmin

Refs: #10

ahmedali8 committed:
feat: add mock minter contract

Mock Minter contract mimicks FriendsAndFamilyMinter and calls subscription contract's 'updateSubscription' function to register the NFT for subscription at the time of mint otherwise the ownerOf funciton of the NFT will revert

Refs: #11

ahmedali8 committed:
chore: temporary disable tests file

sweetmantech commented on contracts/Subscription.sol:
```diff
@@ -109,56 +116,97 @@ contract Subscription is IERC5643 {
      */
     function supportsInterface(bytes4 interfaceId) public view virtual returns (bool) {
-        return interfaceId == type(IERC5643).interfaceId || erc721.supportsInterface(interfaceId);
+        return interfaceId == type(IERC5643).interfaceId || IERC721(cre8orsNFT).supportsInterface(interfaceId);
```
first time seeing `supportsInterface` chained to another contract. 🧑‍🏫 

can you help me understand why our Subscription contract supports all the interfaces for our base Cre8or? learning moment for me. 💡

ahmedali8 replied:
As `ERC5643` is a minimal extension of `ERC721` so it needs to be compliant with `IERC721` as well.

[Cygar's ERC5643](https://github.com/cygaar/ERC5643/blob/55bc2fc9ab3656387bd59eefbcc3871719f6a5eb/src/ERC5643.sol#L170) implementation inherits from `ERC721` so it overrides the `supportsInterface` and such pattern can also be seen in [ens' subscription implementation](https://github.com/ensdomains/ens-contracts/blob/e6b136e979084de3761c125142620304173990ca/contracts/ethregistrar/BaseRegistrarImplementation.sol#L187)

As we are not inheriting our base Cre8ors but we are attaching a `ERC5643` which needs to be compliant that 's why this line.

Let me know what are your thoughts @SweetmanTech

ahmedali8 committed:
refactor: modularize ERC5643 abstract contract

ahmedali8 committed:
feat: add errors
style: add sections for readability

ahmedali8 committed:
feat: modularize Subscription contract
style: add sections
feat: add function overloads for subscription with mint variants
docs: add natspec

ahmedali8 committed:
chore: eslint ignore lib folder
chore: solhint 'no-empty-blocks warning' off

ahmedali8 committed:
docs: add NATSPEC

ahmedali8 committed:
ci: add slither ci/cd github action

Ref: #13

ahmedali8 committed:
chore: filter sol_script and sol_test from slither

ahmedali8 committed:
feat: MockNFT inherits from IMockNFT
fix: add zero-address check

ahmedali8 committed:
fix: TODO price calculation and validation

ahmedali8 committed:
ci: update slither action

ahmedali8 committed:
ci(slither): fix node version and target

sweetmantech commented on contracts/Subscription.sol:
```diff
@@ -1,205 +1,356 @@
+    /// @dev Initializes the Subscription contract with the address of the NFT and the minter contract addresses.
+    /// @param _cre8orsNFT The address of the NFT contract. (cre8ors)
+    /// @param _minter The address of the minter. (FriendsAndFamilyMinter)
+    constructor(address _cre8orsNFT, address _minter) ERC5643(_cre8orsNFT, _minter) { }
```
remove `minter` prop and instead use `MINTER` role on base cre8ors.

ahmedali8 replied:
Done!

sweetmantech commented on contracts/abstracts/Minter.sol:
```diff
@@ -0,0 +1,45 @@
+    address public minter;
+
+    /// @dev Modifier to restrict access to only the minter.
+    modifier onlyMinter() {
+        if (msg.sender != minter) {
+            revert Access_OnlyMinter();
+        }
+        _;
+    }
```
instead of storing a minter in storage, can you please use the `MINTER_ROLE` found in base `cre8ors`?

you can follow the pattern found in `src/cre8ing.sol`
<img width="830" alt="Screenshot 2023-08-02 at 1 15 56 AM" src="https://github.com/defientco/crea8ors-subscription-contracts/assets/23249402/c32dd4c6-8145-4220-b443-448193e419f4">

ahmedali8 replied:
Done!

sweetmantech commented on contracts/abstracts/Minter.sol:
```diff
@@ -0,0 +1,45 @@
+    /// @param _cre8orsNFT The address of the Cre8ors NFT contract.
+    /// @param _minter The address of the initial minter.
+    constructor(address _cre8orsNFT, address _minter) Admin(_cre8orsNFT) {
+        minter = _minter;
+
+        emit SetMinter(address(0), _minter);
```
remove this to save gas on setup

ahmedali8 replied:
Done!

sweetmantech commented on contracts/Subscription.sol:
```diff
@@ -1,205 +1,356 @@
+    //////////////////////////////////////////////////////////////*/
+
+    /// @inheritdoc ISubscription
+    function renewSubscriptionWithERC20Payment(
+        uint256 tokenId,
+        uint64 duration,
+        address erc20
+    )
+        external
+        payable
+        override
+    {
+        _validateCallerAsOwnerOrApproved({ caller: msg.sender, tokenId: tokenId });
+
+        _handleERC20PaymentSubscriptionUpdate(tokenId, duration, erc20);
+    }
```
I think we should remove ERC20 methods for simpler contracts & less expensive deployment costs.

what do you think?

ahmedali8 replied:
Yeah, created a branch containing erc20 payment logic for future reference. Removed from current implementation!

sweetmantech commented on contracts/Subscription.sol:
```diff
@@ -1,205 +1,356 @@
+                ++i;
+            }
         }
+    }
```
nice function overload 💯

sweetmantech commented on contracts/Subscription.sol:
```diff
@@ -1,205 +1,356 @@
-        if (msg.value < _getRenewalPrice(tokenId, duration)) {
-            revert InsufficientPayment();
+    /// @inheritdoc ISubscription
+    function updateSubscriptionForFree(
+        uint256[] calldata tokenIds,
+        uint64[] calldata durations
+    )
+        external
+        override
+        onlyMinter
+    {
+        if (tokenIds.length != durations.length) revert LengthMismatch();
+
+        uint256 tokenId;
+        uint64 duration;
+
+        for (uint256 i = 0; i < tokenIds.length;) {
+            tokenId = tokenIds[i];
+            duration = durations[i];
+
+            _validateDurationBetweenMinAndMax(duration);
+            _updateSubscriptionExpiration(tokenId, duration);
+
+            unchecked {
+                ++i;
+            }
         }
+    }
```
we shouldn't need this function for array of `duration[]`, duration should be constant, even if an array of tokenIds is passed.

`duration` will always be `365 days` for mint day context on 8/8 mint day

ahmedali8 replied:
Done!

sweetmantech commented on contracts/Subscription.sol:
```diff
@@ -1,205 +1,356 @@
+
+        ERC20TransferHelper.safeTransferFrom(IERC20(erc20), msgSender, addressThis, tokensQtyToTransfer);
     }
 }
```
what function does an owner of a cre8or call to extend their subscription?

Looks like all the methods are `minterOnly`

Can we remove all unused methods for 8/8 mint day and focus on MVP functionality? It's cool you're thinking through all the bells & whistles but let's focus on keeping this simple.

ahmedali8 replied:
Changed to `onlyRoleOrAdmin(target, MINTER_ROLE)`

sweetmantech commented on contracts/abstracts/Admin.sol:
```diff
@@ -0,0 +1,26 @@
+/// @notice An abstract contract that restricts access to only the cre8orsNFT contract's admin.
+abstract contract Admin {
+    ///@notice The address of the collection contract that mints and manages the tokens.
+    address public cre8orsNFT;
```
`optional` personally, I wouldn't store `cre8orsNFT` on the contract at all. Instead, I would follow how we're using a `target` param in `cre8ing`.

- less redundant storage.
- more reusability across the ecosystem.

<img width="810" alt="Screenshot 2023-08-02 at 1 27 45 AM" src="https://github.com/defientco/crea8ors-subscription-contracts/assets/23249402/a65c024e-646a-4213-a433-857f3447f2a9">

ahmedali8 replied:
Done!

sweetmantech replied:
you're still storing `cre8orsNFT` in contract storage right?
- expected: architecture does not require cre8orsNFT to be stored in contract storage


Am I missing something?
<img width="487" alt="Screenshot 2023-08-02 at 2 46 37 PM" src="https://github.com/defientco/crea8ors-subscription-contracts/assets/23249402/bc90afac-8e40-40eb-b57f-498dc1d92473">

sweetmantech commented on contracts/abstracts/ERC5643.sol:
```diff
@@ -0,0 +1,182 @@
+    /// @dev Constructor for the ERC5643 contract.
+    /// @param _cre8orsNFT The address of the Cre8ors NFT contract.
+    /// @param _minter The address of the minter e.g. FriendsAndFamilyMinter.
+    constructor(address _cre8orsNFT, address _minter) PaymentSystem(_cre8orsNFT, _minter) { }
```
`optional` ideal architecture does not need a `constructor`
- `target` param replaces need for `_cre8orsNFT`
-  `onlyRoleOrAdmin(MINTER_ROLE)`  modifier  replaces need for `_minter`

sweetmantech commented on contracts/abstracts/PaymentSystem.sol:
```diff
@@ -0,0 +1,81 @@
+import { Minter } from "./Minter.sol";
+
+abstract contract PaymentSystem is Minter {
+    struct ERC20State {
+        bool isActive;
+        uint256 chargeAmount; // pricePerSecond
+    }
```
can you please remove usage of ERC20?

we have no plans to use ERC20 for subscriptions, only native ETH

ahmedali8 replied:
Only native ETH now

sweetmantech commented on contracts/abstracts/PaymentSystem.sol:
```diff
@@ -0,0 +1,81 @@
+    error ETHTransferFailed();
+
+    // solhint-disable-next-line no-empty-blocks
+    constructor(address _crea8orsNFT, address _minter) Minter(_crea8orsNFT, _minter) { }
```
preferred architecture does not need `constructors` for `cre8ors` or `minters`

ahmedali8 replied:
PaymentSystem would now require only `pricePerSecond` to be set via constructor.

sweetmantech commented on contracts/mocks/MockNFT.sol:
```diff
@@ -0,0 +1,60 @@
+
+    // REQUIRED
+    // override ownerOf
+    function ownerOf(uint256 tokenId) public view override returns (address) {
+        // external call to subscription if it is enabled and present
+        if (isSubscriptionEnabled && subscription != address(0)) {
+            // handles revert for us
+            ISubscription(subscription).validateSubscription(tokenId);
+        }
+        return super.ownerOf(tokenId);
+    }
```
in this architecture, how can we send expired NFTs to:
- `treasury.cre8ors.eth`
- `burn` address
- etc?

looks like we only have the ability to `revert` right?

ahmedali8 replied:
`ISubscription(subscription).isSubscriptionValid(tokenId)` can be used instead of `ISubscription(subscription).validateSubscription(tokenId)` which returns `bool` without reverting to handle the flow in the `ownerOf override`.

You need to have a `burn` function which the admin can access when NFT is expired. This `ownerOf` override will have no effect on the burning mechanism of `cre8ors` NFT as it is inheriting from `ERC721A` [`_burn`](https://github.com/chiru-labs/ERC721A/blob/c66792253580acb2b7a873046858175d213e5b65/contracts/ERC721A.sol#L951)

sweetmantech replied:
we should be able to return any arbitrary value for `ownerOf`

core requirement: if subscription expires, we can change the `ownerOf` response to be `address(0)` or any other address.

ahmedali8 replied:
https://github.com/defientco/crea8ors-subscription-contracts/blob/274d94f75a005ff8f671ee662217e289be03d6e3/contracts/mocks/MockNFT.sol#L39

ahmedali8 committed:
refactor: remove cre8ors storage variable and use target param

less redundant storage
more reusability

ahmedali8 committed:
refactor: remove Minter contract
feat: add onlyRoleOrAdmin modifier in Admin abstract contract

ahmedali8 committed:
refactor: remove erc20 payments
doc: add natspec

ahmedali8 committed:
fix: withdraw and natspec

ahmedali8 commented on contracts/abstracts/ERC5643.sol:
```diff
@@ -0,0 +1,182 @@
+    //////////////////////////////////////////////////////////////*/
+
+    /// @inheritdoc IERC5643
+    function renewSubscription(uint256 tokenId, uint64 duration) external payable virtual override {
```
@SweetmanTech we can't change the function definition of `renewSubscription` to pass it on to `_isApprovedOrOwner` so we still need the `cre8orsNFT` state variable

ahmedali8 commented on contracts/abstracts/ERC5643.sol:
```diff
@@ -0,0 +1,182 @@
+    }
+
+    // if crea8ors has an `isApprovedOrOwner` then this can be removed
+    function _isApprovedOrOwner(address spender, uint256 tokenId) internal view virtual returns (bool) {
```
@SweetmanTech this function requires `cre8orsNFT` state variable

ahmedali8 committed:
chore: use nomic solidity extension as Juan causing issues

ahmedali8 committed:
feat: add dev natspec tag
feat: add duration error

ahmedali8 committed:
feat: set pricePerSecond via constructor

ahmedali8 committed:
refactor: move validation internal functions to modifiers
refactor: remove erc20 functions
feat: update minRenewalDuration via constructor

refactor: update constructor

ahmedali8 committed:
refactor: remove erc20 functions
feat: add target mechanism
perf: function params use two slots
refactor: remove durations[] implementation functions

ahmedali8 committed:
refactor: remove erc20 payment functions
feat: add target mechanism for onlyAdmin and onlyRoleOrAdmin(MINTER_ROLE)
refactor: add validation modifiers instead of internal functions
refactor: remove tokenIds[] and durations[] variants of subscription
refactor: rename nativeCurrencyPrice to pricePerSecond

ahmedali8 committed:
refactor: use target implementation
refactor: use updateSubscriptionForFree

ahmedali8 committed:
perf: remove unused imports

ahmedali8 committed:
feat: write deploy scripts

sweetmantech commented on contracts/Subscription.sol:
```diff
@@ -1,205 +1,181 @@
-        // check data length
-        if (input.length == 0) revert CalledWithOutData();
+    constructor(
+        address cre8orsNFT_,
```
should not need to store `cre8orsNFT_` in storage

sweetmantech commented on contracts/abstracts/ERC5643.sol:
```diff
@@ -0,0 +1,200 @@
+    uint64 public maxRenewalDuration; // 0 value means lifetime extension
+
+    ///@notice The address of the collection contract that mints and manages the tokens.
+    address public cre8orsNFT;
```
should not need to store this on the contract.

ahmedali8 committed:
feat: add use-case if subscription expires, we can change the ownerOf response to be address(0) or any other address

sweetmantech reviewed (approved):
ship it 🚢

ahmedali8 committed:
feat: add setter function for cre8orsNFT

ahmedali8 committed:
chore: ignore deployments

ahmedali8 committed:
feat: add a base contract for common functionality

ahmedali8 committed:
feat: Admin inherits Base

ahmedali8 committed:
doc: add natspec
feat: add zero address validation

ahmedali8 committed:
style: add section

ahmedali8 committed:
feat: update error
docs: add natspec

ahmedali8 committed:
feat: add events
refactor: change validateSubscription to checkSubscription

ahmedali8 committed:
refactor: update mocks

ahmedali8 committed:
docs: add natspec
feat: add events
refactor: follow interface changes

ahmedali8 committed:
refactor: remove libraries

ahmedali8 committed:
chore: lint ts code

ahmedali8 committed:
test: update basic integration test flow

ahmedali8 committed:
feat: deploy subscription
